zkevm-circuits/src/tx_circuit.rs

```diff
@@ -0,0 +1,542 @@
+//! The transaction circuit implementation.
+
+// Naming notes:
+// - *_be: Big-Endian bytes
+// - *_le: Little-Endian bytes
+
+mod sign_verify;
+
+use crate::util::Expr;
+use eth_types::{Address, Bytes, Field, ToBigEndian, ToLittleEndian, ToScalar, Word, U256};
+use ff::PrimeField;
+use group::GroupEncoding;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter, Region, SimpleFloorPlanner},
+    plonk::{Advice, Circuit, Column, ConstraintSystem, Error},
+    poly::Rotation,
+};
+use itertools::Itertools;
+use lazy_static::lazy_static;
+use libsecp256k1;
+use log::error;
+use num::Integer;
+use num_bigint::BigUint;
+use rlp::RlpStream;
+use secp256k1::Secp256k1Affine;
+use sha3::{Digest, Keccak256};
+use sign_verify::{SignData, SignVerifyChip, SignVerifyConfig};
+pub use sign_verify::{POW_RAND_SIZE, VERIF_HEIGHT};
+use std::convert::TryInto;
+use std::marker::PhantomData;
+use subtle::CtOption;
+
+lazy_static! {
+    // Scalar
+    static ref SECP256K1_Q: BigUint = BigUint::from_slice(&[
+        0xd0364141, 0xbfd25e8c,
+        0xaf48a03b, 0xbaaedce6,
+        0xfffffffe, 0xffffffff,
+        0xffffffff, 0xffffffff,
+    ]);
+}
+
+/// Transaction to be verified by the TxCircuit
+#[derive(Clone, Default, Debug)]
+pub struct Transaction {
```

Comment:
>About a, currently it doesn't have the signature fields, which are required for the tx_circuit. Maybe we can have a unified struct merging a and the tx_circuit struct, where the signature values are 0 when unused. Notice that I'm saying 0 and not None, because if it was an Option the tx_circuit would be using a lot of unwraps and I prefer avoiding that.

Makes sense. I understand the concern here. So I leave it to your call on this regard. I think the less structs we have with the same name, the better. But unwrapping all the time can be a pain.

>I think b and c can be merged into one in the future. The idea of having duplicate structs between b and c was to allow the development of the circuit input builder in parallel with the circuits while minimizing the changes that one implies to the other. But once things stabilize, I think we can merge those.

Don't you think that's stable enough already to do it? I would really like to push (if not in this PR in a follow-up one soon) the merge of the Tx types.

Reply:
> Makes sense. I understand the concern here. So I leave it to your call on this regard. I think the less structs we have with the same name, the better. But unwrapping all the time can be a pain.

In https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/484/commits/c3ba0276312a056b96b79ee312f8f1c9e68b0418 I'm reusing geth_types::Transaction for the tx_circuit.  Thanks for mentioning this because now there's one less repeated struct and it was an easy change :)

> Don't you think that's stable enough already to do it? I would really like to push (if not in this PR in a follow-up one soon) the merge of the Tx types.

I've created an issue https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/528 Feel free to assign it to yourself if you'd like to work on that! :)